Mono.Cecil.Cil/MethodBody.cs

```diff
@@ -163,10 +165,46 @@ protected override void OnSet (VariableDefinition item, int index)
 
 		protected override void OnRemove (VariableDefinition item, int index)
 		{
+			UpdateVariableIndices (index + 1, -1, item);
 			item.index = -1;
+		}
 
-			for (int i = index + 1; i < size; i++)
-				items [i].index = i - 1;
+		void UpdateVariableIndices (int startIndex, int offset, VariableDefinition variableToRemove = null)
+		{
+			for (int i = startIndex; i < size; i++)
+				items [i].index = i + offset;
+
+			var debug_info = method == null ? null : method.debug_info;
+			if (debug_info == null || debug_info.Scope == null)
+				return;
+
+			foreach (var scope in debug_info.GetScopes ()) {
+				if (scope.HasVariables) {
```

Comment:
Please transform this condition into an early continue if there's no variable to reduce nesting.